### PR TITLE
Performance improvements of IOSim

### LINKED ISSUE #3673: Performance review of IOSim

### PR DESCRIPTION
**Pull request by coot**
Resolves #3673.

### CODE REVIEW

File: io-sim/src/Control/Monad/IOSim/Internal.hs [hunk 1]
```diff
@@ -198,7 +198,8 @@ schedule !thread@Thread{
   assert (invariant (Just thread) simstate) $
   case action of
 
-    Return x -> case ctl of
+    Return x -> {-# SCC "schedule.Return" #-}
```
Comment (bolt12) [inline io-sim/src/Control/Monad/IOSim/Internal.hs:201]: Is this necessary to stay? (The same for all changes in this commit 1d5d3763e94b5a16b9e2177b5d40351d5157079b)

File: io-sim/src/Control/Monad/IOSimPOR/Internal.hs [hunk 2]
```diff
@@ -1057,7 +1057,9 @@ execAtomically time tid tlbl nextVid0 action0 k0 =
        -> ST s (SimTrace c)
     go ctl !read !written writtenSeq createdSeq !nextVid action = assert localInvariant $
                                                        case action of
-      ReturnStm x -> case ctl of
+      ReturnStm x ->
+        {-# SCC "execAtomically.go.ReturnStm" #-}
```
Comment (bolt12) [inline io-sim/src/Control/Monad/IOSimPOR/Internal.hs:1061]: Is this necessary to stay? (The same for the rest of this commit 22a4ed1d56c926bf89038f12a699e213f33823a8)

**Review by bolt12 — APPROVED** [review]
LGTM! Just some small things

File: io-sim/src/Control/Monad/IOSim/Internal.hs (hunk 1 above)
Comment (coot) [reply io-sim/src/Control/Monad/IOSim/Internal.hs:201]: It doesn't cost us to keep it, and might prove to be useful in the future.

File: io-sim/src/Control/Monad/IOSimPOR/Internal.hs (hunk 2 above)
Comment (coot) [reply io-sim/src/Control/Monad/IOSimPOR/Internal.hs:1061]: As above.

**Comment by coot** [comment]
bors merge

**Pull request merged**